python/ray/serve/tests/unit/test_application_state.py

```diff
@@ -236,6 +237,44 @@ def test_unhealthy(self, get_deployments_statuses, mocked_application_state):
         assert status == ApplicationStatus.UNHEALTHY
         assert error_msg
 
+    @patch.object(ApplicationState, "get_deployments_statuses")
+    def test_autoscaling(self, get_deployments_statuses, mocked_application_state):
+        app_state, _ = mocked_application_state
+        app_state._status = ApplicationStatus.RUNNING
+        get_deployments_statuses.return_value = [
+            DeploymentStatusInfo(
+                "a", DeploymentStatus.HEALTHY, DeploymentStatusTrigger.DEPLOY
+            ),
+            DeploymentStatusInfo(
+                "b", DeploymentStatus.UPSCALING, DeploymentStatusTrigger.AUTOSCALE
+            ),
+            DeploymentStatusInfo(
+                "c", DeploymentStatus.DOWNSCALING, DeploymentStatusTrigger.AUTOSCALE
+            ),
+        ]
+        status, error_msg = app_state._determine_app_status()
+        assert status == ApplicationStatus.RUNNING
+
+    @patch.object(ApplicationState, "get_deployments_statuses")
+    def test_manual_scale_num_replicas(
+        self, get_deployments_statuses, mocked_application_state
+    ):
+        app_state, _ = mocked_application_state
+        app_state._status = ApplicationStatus.RUNNING
+        get_deployments_statuses.return_value = [
+            DeploymentStatusInfo(
+                "a", DeploymentStatus.HEALTHY, DeploymentStatusTrigger.DEPLOY
+            ),
+            DeploymentStatusInfo(
+                "b", DeploymentStatus.UPSCALING, DeploymentStatusTrigger.CONFIG_UPDATE
+            ),
+            DeploymentStatusInfo(
+                "c", DeploymentStatus.DOWNSCALING, DeploymentStatusTrigger.CONFIG_UPDATE
+            ),
+        ]
+        status, error_msg = app_state._determine_app_status()
```

Comment:
here's a case where having a static pure function would make testing more convenient -- pass in whatever is used to determine the status instead of pulling it from instance variables and you could write this test without mocks and/or reaching into internal fields details